opc-ua-sdk/sdk-server/src/main/java/org/eclipse/milo/opcua/sdk/server/services/DiscoveryServices.java

```diff
@@ -0,0 +1,345 @@
+            .limit(recordCount);
+
+        if (serviceRequest.getRequest().getServerCapabilityFilter() != null &&
+            serviceRequest.getRequest().getServerCapabilityFilter().length > 0) {
+
+            filteredServers =
+                filteredServers.filter(serverOnNetwork -> serverOnNetwork.getServerCapabilities() == null ||
+                    Arrays.asList(serverOnNetwork.getServerCapabilities())
+                        .containsAll(
+                            Arrays.asList(serviceRequest.getRequest().getServerCapabilityFilter())));
+        }
+
+        ResponseHeader header = serviceRequest.createResponseHeader();
+
+        serviceRequest.setResponse(
+            new FindServersOnNetworkResponse(header, new DateTime(mdnsHelper.getLastServerOnNetworkIdReset()),
+                filteredServers.toArray(ServerOnNetwork[]::new)));
+    }
+
+    public StatusCode addMulticastRecord(String name, int port, String path, String[] capabilities) {
+        if (!multicastEnabled) {
+            return new StatusCode(StatusCodes.Bad_ServiceUnsupported);
+        }
+        if (!mdnsHelper.addRecord(name, port, path, capabilities)) {
+            return StatusCode.BAD;
+        }
+        return StatusCode.GOOD;
+    }
+
+    public StatusCode removeMulticastRecord(String name, int port, String path) {
+        if (!multicastEnabled) {
+            return new StatusCode(StatusCodes.Bad_ServiceUnsupported);
+        }
+        if (!mdnsHelper.removeRecord(name, port, path)) {
+            return StatusCode.BAD;
+        }
+        return StatusCode.GOOD;
+    }
+
+    private StatusCode processRegisterServer(RegisteredServer requestServer,
+                                             ExtensionObject[] requestDiscoveryConfiguration,
+                                             ArrayList<StatusCode> configurationResults,
+                                             ArrayList<DiagnosticInfo> diagnosticInfos) {
+
+        // check if server already in list
+        RegisteredServer registeredServer = null;
+
+
+        Optional<RegisteredServer> rs =
+            registeredServers.stream().filter(s -> s.getServerUri().compareTo(requestServer.getServerUri()) == 0)
+                .findFirst();
+        if (rs.isPresent()) {
+            registeredServer = rs.get();
+        }
+
+        // set discovery configuration if it is in request (only RegisterServer2)
+        MdnsDiscoveryConfiguration discoveryConfiguration = null;
+        String mdnsName = null;
+        if (requestDiscoveryConfiguration != null) {
+            configurationResults.clear();
+            configurationResults.ensureCapacity(requestDiscoveryConfiguration.length);
+            for (ExtensionObject e : requestDiscoveryConfiguration) {
+                if (discoveryConfiguration == null &&
+                    e.getEncodingTypeId().equals(Identifiers.MdnsDiscoveryConfiguration)) {
```

Comment:
Does this actually work? Has anything tested code path? 

I would have expected that the encoding id would have to be one of `Identifiers.MdnsDiscoveryConfiguration_Encoding_DefaultBinary` or `Identifiers.MdnsDiscoveryConfiguration_Encoding_DefaultXml` (also accessible as `MdnsDiscoveryConfiguration.BinaryEncodingId` and `MdnsDiscoveryConfiguration.XmlEncodingId`)